odoo/tools/sql.py

```diff
@@ -160,19 +187,22 @@ def __iter__(self):
 
     def join(self, args: Iterable) -> SQL:
         """ Join SQL objects or parameters with ``self`` as a separator. """
+        if not self:
+            return SQL(self, *args)
+        # handle no arguments and make args a list
         args = list(args)
-        # optimizations for special cases
-        if len(args) == 0:
+        size = len(args)
+        if size == 0:
             return SQL()
-        if len(args) == 1:
-            return args[0]
-        if not self.__args:
-            return SQL(self.__code.join("%s" for arg in args), *args)
-        # general case: alternate args with self
-        items = [self] * (len(args) * 2 - 1)
+        if size == 1 and isinstance(sql := args[0], SQL):
```

Comment:
Oh, well done.  I didn't notice that bug in my original code.  I just adapted my branch accordingly.